src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/rewriter/SubqueryAliasRewriter.java

```diff
@@ -0,0 +1,148 @@
+import com.alibaba.druid.sql.ast.expr.SQLIdentifierExpr;
+import com.alibaba.druid.sql.ast.expr.SQLInSubQueryExpr;
+import com.alibaba.druid.sql.ast.statement.SQLExprTableSource;
+import com.alibaba.druid.sql.ast.statement.SQLSelectItem;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.alibaba.druid.sql.dialect.mysql.visitor.MySqlASTVisitorAdapter;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * Add the alias for identifier the subquery query.
+ * Use the table alias if it already has one, Auto generate if it doesn't has one.
+ */
+public class SubqueryAliasRewriter extends MySqlASTVisitorAdapter {
+
+    private Deque<Table> tableScope =  new ArrayDeque<>();
+
+    private int aliasSuffix = 0;
+
+    private final static String DOT = ".";
+
+    @Override
+    public boolean visit(MySqlSelectQueryBlock query) {
+        final SQLExprTableSource from = (SQLExprTableSource) query.getFrom();
+        String tableName = from.getExpr().toString().replaceAll(" ", "");
+
+        if (from.getAlias() != null) {
+            tableScope.push(new Table(tableName, from.getAlias()));
+        } else {
+            from.setAlias(createAlias(tableName));
+            tableScope.push(new Table(tableName, from.getAlias()));
+        }
+
+        for (SQLSelectItem item : query.getSelectList()) {
+            final SQLExpr itemExpr = item.getExpr();
+            if (itemExpr instanceof SQLIdentifierExpr) {
+                rewrite(tableScope.peek(), (SQLIdentifierExpr)itemExpr);
+            } else if (itemExpr instanceof SQLAllColumnExpr){
+                item.setExpr(createIdentifierExpr(tableScope.peek(), "*"));
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public void endVisit(MySqlSelectQueryBlock query) {
+        tableScope.pop();
+    }
+
+    @Override
+    public void endVisit(SQLBinaryOpExpr expr) {
+        // ignore the es table name, index/account
+        if (expr.getOperator() == SQLBinaryOperator.Divide && expr.getParent() instanceof SQLExprTableSource) {
+            return;
+        }
+
+        if (expr.getLeft() instanceof SQLIdentifierExpr) {
+            rewrite(tableScope.peek(), (SQLIdentifierExpr)expr.getLeft());
+        }
+        if (expr.getRight() instanceof SQLIdentifierExpr) {
+            rewrite(tableScope.peek(), (SQLIdentifierExpr)expr.getRight());
+        }
+    }
+
+    @Override
+    public void endVisit(SQLInSubQueryExpr expr) {
+        if (expr.getExpr() instanceof SQLIdentifierExpr) {
+            rewrite(tableScope.peek(), (SQLIdentifierExpr)expr.getExpr());
+        }
+    }
+
+    private void rewrite(Table table, SQLIdentifierExpr expr) {
+        final String tableAlias = table.getAlias();
+        final String tableName = table.getName();
+
+        final String exprName = expr.getName();
+        if (exprName.startsWith(tableName + DOT) || exprName.startsWith(tableAlias + DOT)) {
```

Comment:
Need a little clarification. So even if it starts with "`tableAlias.`", we replace "`tableName.`" with "`tableAlias.`"?

Reply:
Yes.